pkg/flow/internal/controller/custom_component_registry.go

```diff
@@ -0,0 +1,93 @@
+// Imported definitions are stored inside of the corresponding import registry.
+type CustomComponentRegistry struct {
+	parent   *CustomComponentRegistry // nil if root config
+	declares map[string]ast.Body      // customComponentName: template
+
+	mut     sync.RWMutex
+	imports map[string]*CustomComponentRegistry // importNamespace: importScope
+}
+
+// NewCustomComponentRegistry creates a new CustomComponentRegistry with a parent.
+// parent can be nil.
+func NewCustomComponentRegistry(parent *CustomComponentRegistry) *CustomComponentRegistry {
+	return &CustomComponentRegistry{
+		parent:   parent,
+		declares: make(map[string]ast.Body),
+		imports:  make(map[string]*CustomComponentRegistry),
+	}
+}
+
+// registerDeclare stores a local declare block.
+func (s *CustomComponentRegistry) registerDeclare(declare *ast.BlockStmt) {
+	s.declares[declare.Label] = declare.Body
+}
+
+// registerImport stores the import namespace.
+// The content will be added later during evaluation.
+// It's important to register it before populating the component nodes
+// (else we don't know which one exists).
+func (s *CustomComponentRegistry) registerImport(importNamespace string) {
+	s.imports[importNamespace] = nil
+}
+
+// updateImportContent updates the content of a registered import.
+// The content of an import node can contain other import blocks.
+// These are considered as "children" of the root import node.
+// Each child has its own CustomComponentRegistry which needs to be updated.
+func (s *CustomComponentRegistry) updateImportContent(importNode *ImportConfigNode) {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	if _, exist := s.imports[importNode.label]; !exist {
+		panic(fmt.Errorf("import %q was not registered", importNode.label))
+	}
+	importScope := NewCustomComponentRegistry(nil)
+	importScope.declares = importNode.ImportedDeclares()
+	importScope.updateImportContentChildren(importNode)
+	s.imports[importNode.label] = importScope
+}
+
+// updateImportContentChildren recurse through the children of an import node
+// and update their scope with the imported declare blocks.
+func (s *CustomComponentRegistry) updateImportContentChildren(importNode *ImportConfigNode) {
+	for _, child := range importNode.ImportConfigNodesChildren() {
+		childScope := NewCustomComponentRegistry(nil)
+		childScope.declares = child.ImportedDeclares()
+		childScope.updateImportContentChildren(child)
+		s.imports[child.label] = childScope
+	}
+}
+
+// deepCopy returns a deep copy of the full scope (including parents and imports).
+func (s *CustomComponentRegistry) deepCopy() *CustomComponentRegistry {
+	s.mut.Lock()
+	defer s.mut.Unlock()
+	newScope := NewCustomComponentRegistry(nil)
+
+	if s.parent != nil {
+		newScope.parent = s.parent.deepCopy()
+	}
+
+	for k, v := range s.declares {
+		if v != nil {
+			newScope.declares[k] = v
+		}
+	}
+
+	for k, v := range s.imports {
+		if v != nil {
+			newScope.imports[k] = v.deepCopy()
+		}
+	}
+	return newScope
+}
```

Comment:
Why do we need to create a deep copy? Code for producing copies is prone to become stale as new functionality is added, leading to obscure bugs. If possible, I recommend avoiding the need to make copies as much as possible.

Reply:
My concern is that because we share the registry through all modules, a parent may modify it while the child is accessing it. I'm not sure if the mutex is enough in this case because when it will be released by the parent, the child might not find the content that it was looking for. The change will trigger a reload of the child, but you have this small window where the behavior of the child is undefined (may not find the content or may find a content that does not work with the current state of the child). I remove it now because I think that such situation should occur only with imports (with declare only we would create a new pointer on reload). What do you think?

Reply:
Let's leave it for now (unless you've already removed it), I'd like to avoid any larger changes, especially if you think they might be brought right back